waveformseekbar: Enhance the redraw performance and frequence

ptitjes opened the pull request:
I carefully enhanced the performance by using Cairo's clipping mechanism. Now, only the minimum required region of the waveform is queued for redraw.

Then the interval time between the redraws is computed depending on the width of the waveform and the length of the song. With a 2'50 song rendered on half of a MacBookPro 13", the redraw happens every ~250 milliseconds (4 times more than currently). On a full screen MacBookPro 13" it happens every ~90 milliseconds. Due to the careful handling of the clip rectangle, the occupation of the CPU shows no real difference.

lazka commented on quodlibet/quodlibet/qltk/tracker.py:
```diff
@@ -69,7 +82,7 @@ def __paused(self, player, paused):
             self.__stop = False
             if self.__id is None:
-                self.__id = GLib.timeout_add_seconds(1, self.__update)
+                self.__id = GLib.timeout_add(self.__interval, self.__update)
```
TimeTracker uses timeout_add_seconds() because that gets synced with all other timeout_add_seconds() calls (also from other apps) and allows us to wake up only once per second. 

So in case it is 1000 it should still call timeout_add_seconds().

ptitjes commented:
@lazka OK, I fixed the reuse of the existing waking up every seconds. Also, I fixed two bugs in my code.
I think it is good now, if you want to review my additions. Thanks.

lazka commented:
Thanks. Just one thing (I haven't looked at the drawing part), the invalidated area looks like its width is based on the elapsed time, is that on purpose?

![out](https://cloud.githubusercontent.com/assets/991986/23549555/1064813c-000d-11e7-9db9-293f757d0a5c.gif)

ptitjes commented:
@lazka Thanks, was a big bug ! :)
BTW, how do you highlight the invalidated areas like that ? Can't find that on my search engine...

ptitjes commented:
OK, finally found it: `GTK_DEBUG=updates` in the environment.
So I confirm it now works as expected.

ptitjes commented:
There are still some full redraws that could be avoided when seeking through the song.
But still it is far better than what it was before.
I will enhance the redraw behaviours on seeking later.

ptitjes commented:
Well later is now...

![redraws2](https://cloud.githubusercontent.com/assets/64956/23553802/a1b8811e-0022-11e7-9832-f533a74391b4.gif)

lazka commented:
:+1:

lazka commented:
If you press "ctrl+shift+i" while running QL, the gtk inspector will start where you can also enable visual updates etc..

ptitjes commented:
"ctrl+shift+i" doesn't trigger the inspector here !

ptitjes commented:
OK, to have Ctrl-Shift-D and Ctrl-Shift-I toggle the inspector, I needed that first:
```
gsettings set org.gtk.Settings.Debug enable-inspector-keybinding true
```

ptitjes commented:
After some more testing it seems to me that maybe one of `_song_started` or `_song_changed` could be ignored. I don't know which one however as I am not sure to fully understand the difference.

Other than that, I'm quite happy with the result.

lazka commented:
> After some more testing it seems to me that maybe one of _song_started or _song_changed could be ignored. I don't know which one however as I am not sure to fully understand the difference.

song-started is after app.player.song changes, song-changed is after any tags of the provided songs have changed.

ptitjes commented:
> song-started is after app.player.song changes, song-changed is after any tags of the provided songs have changed.

Yeah! I finally figured out as you can see in my last commit :)

ptitjes commented:
Thank you!
